Incomplete shortcodes crash hugo running in serve mode.

<!-- Please answer these questions before submitting your issue. Thanks! -->

### What version of Hugo are you using (`hugo version`)?

<pre>
$ hugo version

Hugo Static Site Generator v0.64.0/extended darwin/amd64 BuildDate: unknown
</pre>

### Does this issue reproduce with the latest release?

Yes

---

I think this is likely the same as https://github.com/gohugoio/hugo/issues/6857

Steps to reproduce:

(Mac)
Run Visual Studio Code with auto-save enabled (set at 250ms).
Run `hugo serve -D -F`

If you type 

```{{< >}}```

or something that can be parsed as a shortcode, but isn't yet valid, hugo seems to produce the stack dump below.

Fix:  gracefully handle bad shortcodes, and just log an error rather than crash.

```
Source changed "/dev/mysite/mermaid.md": WRITE
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x4b62b10]

goroutine 980 [running]:
github.com/gohugoio/hugo/hugolib.(*shortcodeHandler).extractShortcode(0xc00220cb80, 0x4, 0x0, 0xc0027686e0, 0x1, 0x1, 0x5554460)
        /private/tmp/hugo-20200204-77776-1t93oty/hugo-0.64.0/src/github.com/gohugoio/hugo/hugolib/shortcode.go:499 +0x550
github.com/gohugoio/hugo/hugolib.(*pageState).mapContent(0xc002734f30, 0xc00259c2a0, 0xc002268280, 0x0, 0xc000588b38)
        /private/tmp/hugo-20200204-77776-1t93oty/hugo-0.64.0/src/github.com/gohugoio/hugo/hugolib/page.go:738 +0xb16
github.com/gohugoio/hugo/hugolib.newPageWithContent.func1(0xc00259c2a0, 0x402f84a, 0xc000588b90)
        /private/tmp/hugo-20200204-77776-1t93oty/hugo-0.64.0/src/github.com/gohugoio/hugo/hugolib/page__new.go:233 +0x4b
github.com/gohugoio/hugo/hugolib.(*pagesMap).initPageMeta.func1()
        /private/tmp/hugo-20200204-77776-1t93oty/hugo-0.64.0/src/github.com/gohugoio/hugo/hugolib/pages_map.go:77 +0x50
sync.(*Once).doSlow(0xc000030020, 0xc000588c08)
        /usr/local/Cellar/go/1.13.7/libexec/src/sync/once.go:66 +0xe3
sync.(*Once).Do(...)
        /usr/local/Cellar/go/1.13.7/libexec/src/sync/once.go:57
github.com/gohugoio/hugo/hugolib.(*pagesMap).initPageMeta(0xc002768530, 0xc002734f30, 0xc00259c2a0, 0x58e463a, 0x4)
        /private/tmp/hugo-20200204-77776-1t93oty/hugo-0.64.0/src/github.com/gohugoio/hugo/hugolib/pages_map.go:75 +0xa1
github.com/gohugoio/hugo/hugolib.(*pagesMap).initPageMetaFor(0xc002768530, 0xc0023fc156, 0x6, 0xc00259c2a0, 0x0, 0xc00271e1d0)
        /private/tmp/hugo-20200204-77776-1t93oty/hugo-0.64.0/src/github.com/gohugoio/hugo/hugolib/pages_map.go:95 +0x429
github.com/gohugoio/hugo/hugolib.(*pagesMap).assemblePageMeta.func1(0xc0023fc156, 0x6, 0x56cb480, 0xc00259c2a0, 0xc00259c1c0)
        /private/tmp/hugo-20200204-77776-1t93oty/hugo-0.64.0/src/github.com/gohugoio/hugo/hugolib/pages_map.go:199 +0x66
github.com/armon/go-radix.recursiveWalk(0xc002735320, 0xc00057fea0, 0x56cb480)
        /Users/brew/Library/Caches/Homebrew/go_cache/pkg/mod/github.com/armon/go-radix@v1.0.0/radix.go:519 +0xf0
github.com/armon/go-radix.recursiveWalk(0xc002735410, 0xc00057fea0, 0x56cb400)
        /Users/brew/Library/Caches/Homebrew/go_cache/pkg/mod/github.com/armon/go-radix@v1.0.0/radix.go:525 +0x7e
github.com/armon/go-radix.recursiveWalk(0xc002735140, 0xc00057fea0, 0xc0007ef110)
        /Users/brew/Library/Caches/Homebrew/go_cache/pkg/mod/github.com/armon/go-radix@v1.0.0/radix.go:525 +0x7e
github.com/armon/go-radix.recursiveWalk(0xc002735050, 0xc00057fea0, 0x0)
        /Users/brew/Library/Caches/Homebrew/go_cache/pkg/mod/github.com/armon/go-radix@v1.0.0/radix.go:525 +0x7e
github.com/armon/go-radix.(*Tree).Walk(...)
        /Users/brew/Library/Caches/Homebrew/go_cache/pkg/mod/github.com/armon/go-radix@v1.0.0/radix.go:447
github.com/gohugoio/hugo/hugolib.(*pagesMap).assemblePageMeta(0xc002768530, 0xc0003f0a80, 0x0)
        /private/tmp/hugo-20200204-77776-1t93oty/hugo-0.64.0/src/github.com/gohugoio/hugo/hugolib/pages_map.go:196 +0x6c
github.com/gohugoio/hugo/hugolib.(*HugoSites).assemble.func1(0x0, 0x0)
        /private/tmp/hugo-20200204-77776-1t93oty/hugo-0.64.0/src/github.com/gohugoio/hugo/hugolib/hugo_sites_build.go:266 +0xff
golang.org/x/sync/errgroup.(*Group).Go.func1(0xc002734ff0, 0xc002735020)
        /Users/brew/Library/Caches/Homebrew/go_cache/pkg/mod/golang.org/x/sync@v0.0.0-20190911185100-cd5d95a43a6e/errgroup/errgroup.go:57 +0x64
created by golang.org/x/sync/errgroup.(*Group).Go
        /Users/brew/Library/Caches/Homebrew/go_cache/pkg/mod/golang.org/x/sync@v0.0.0-20190911185100-cd5d95a43a6e/errgroup/errgroup.go:54 +0x66
```

--- a/hugolib/shortcode.go
+++ b/hugolib/shortcode.go
@@ -33,8 +33,6 @@ import (
 	"github.com/gohugoio/hugo/parser/pageparser"
 	"github.com/gohugoio/hugo/resources/page"
 
-	_errors "github.com/pkg/errors"
-
 	"github.com/gohugoio/hugo/common/maps"
 	"github.com/gohugoio/hugo/common/text"
 	"github.com/gohugoio/hugo/common/urls"
@@ -310,7 +308,7 @@ func renderShortcode(
 			var found bool
 			tmpl, found = s.TextTmpl().Lookup(templName)
 			if !found {
-				return "", false, _errors.Errorf("no earlier definition of shortcode %q found", sc.name)
+				return "", false, errors.Errorf("no earlier definition of shortcode %q found", sc.name)
 			}
 		}
 	} else {
@@ -417,7 +415,7 @@ func (s *shortcodeHandler) renderShortcodesForPage(p *pageState, f output.Format
 	for _, v := range s.shortcodes {
 		s, more, err := renderShortcode(0, s.s, tplVariants, v, nil, p)
 		if err != nil {
-			err = p.parseError(_errors.Wrapf(err, "failed to render shortcode %q", v.name), p.source.parsed.Input(), v.pos)
+			err = p.parseError(errors.Wrapf(err, "failed to render shortcode %q", v.name), p.source.parsed.Input(), v.pos)
 			return nil, false, err
 		}
 		hasVariants = hasVariants || more
@@ -460,6 +458,10 @@ Loop:
 		switch {
 		case currItem.IsLeftShortcodeDelim():
 			next := pt.Peek()
+			if next.IsRightShortcodeDelim() {
+				// no name: {{< >}} or {{% %}}
+				return sc, errors.New("Shortcode has no name")
+			}
 			if next.IsShortcodeClose() {
 				continue
 			}
@@ -506,7 +508,7 @@ Loop:
 						// return that error, more specific
 						continue
 					}
-					return sc, fail(_errors.Errorf("shortcode %q has no .Inner, yet a closing tag was provided", next.Val), next)
+					return sc, fail(errors.Errorf("shortcode %q has no .Inner, yet a closing tag was provided", next.Val), next)
 				}
 			}
 			if next.IsRightShortcodeDelim() {
@@ -536,7 +538,7 @@ Loop:
 			// Used to check if the template expects inner content.
 			templs := s.s.Tmpl().LookupVariants(sc.name)
 			if templs == nil {
-				return nil, _errors.Errorf("template for shortcode %q not found", sc.name)
+				return nil, errors.Errorf("template for shortcode %q not found", sc.name)
 			}
 
 			sc.info = templs[0].(tpl.Info)
@@ -637,7 +639,7 @@ func renderShortcodeWithPage(h tpl.TemplateHandler, tmpl tpl.Template, data *Sho
 
 	err := h.Execute(tmpl, buffer, data)
 	if err != nil {
-		return "", _errors.Wrap(err, "failed to process shortcode")
+		return "", errors.Wrap(err, "failed to process shortcode")
 	}
 	return buffer.String(), nil
 }

--- a/hugolib/shortcode.go
+++ b/hugolib/shortcode.go
@@ -460,7 +460,7 @@ Loop:
 			next := pt.Peek()
 			if next.IsRightShortcodeDelim() {
 				// no name: {{< >}} or {{% %}}
-				return sc, errors.New("Shortcode has no name")
+				return sc, errors.New("shortcode has no name")
 			}
 			if next.IsShortcodeClose() {
 				continue

--- a/hugolib/shortcode_test.go
+++ b/hugolib/shortcode_test.go
@@ -345,7 +345,7 @@ func TestShortcodeWrappedInPIssue(t *testing.T) {
 // #6866
 func TestShortcodeIncomplete(t *testing.T) {
 	t.Parallel()
-	CheckShortCodeMatchAndError(t, `{{<          >}}`, ".*Shortcode has no name.*", nil, true)
+	CheckShortCodeMatchAndError(t, `{{<          >}}`, ".*shortcode has no name.*", nil, true)
 }
 
 func TestExtractShortcodes(t *testing.T) {

--- a/hugolib/shortcode_test.go
+++ b/hugolib/shortcode_test.go
@@ -65,8 +65,9 @@ title: "Title"
 		t.Fatalf("Shortcode rendered error %s.", err)
 	}
 
-	if err == nil && expectError {
-		t.Fatalf("No error from shortcode")
+	if expectError {
+		c.Assert(err, qt.ErrorMatches, expected)
+		return
 	}
 
 	h := b.H
@@ -339,6 +340,12 @@ func TestShortcodeWrappedInPIssue(t *testing.T) {
 
 {{< bug >}}
 `, "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\n\nxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx", wt)
+}
+
+// #6866
+func TestShortcodeIncomplete(t *testing.T) {
+	t.Parallel()
+	CheckShortCodeMatchAndError(t, `{{<          >}}`, ".*Shortcode has no name.*", nil, true)
 }
 
 func TestExtractShortcodes(t *testing.T) {

